### PR TITLE
Console log

### PR DESCRIPTION
**Pull request by jonacollazo**
Helps on this issue: https://github.com/oven-sh/bun/issues/188
One big difference with node is the amount of stuff printed out, bun prints what node would with JSON.stringify
Code snippet used to quickly compare
```zig
pub const max_u16_ascii = @splat(8, @as(u16, 127));
pub const min_u16_ascii = @splat(8, @as(u16, 0x20));

pub fn toUTF16Literal(comptime str: []const u8) []const u16 {
    comptime {
        comptime var output: [str.len]u16 = undefined;

        for (str) |c, i| {
            output[i] = c;
        }

        const Static = struct {
            pub const literal: []const u16 = output[0..];
        };

        return Static.literal;
    }
}

pub fn simd_double_loop(slice: []const u16) ?u32 {
    const VECTOR_SIZE:usize = 8;
    const remainder:usize = slice.len % VECTOR_SIZE;
    
    const remaining = if (remainder != 0) slice[0..slice.len - remainder] else slice;

    var indices: @Vector(VECTOR_SIZE, usize) = std.simd.iota(usize, VECTOR_SIZE);
    const increment:  @Vector(VECTOR_SIZE, usize) =  @splat(VECTOR_SIZE, VECTOR_SIZE);
    var i: usize = 0;
    while (i < remaining.len) : (i += VECTOR_SIZE) {
        const mins: @Vector(VECTOR_SIZE, u16) = remaining[i..][0..VECTOR_SIZE].*;
        const mask: @Vector(VECTOR_SIZE, bool) = @select(bool, mins < min_u16_ascii, @splat(VECTOR_SIZE, true), mins > max_u16_ascii); 
        const min_indices  = @select(usize,  mask, indices, @splat(VECTOR_SIZE, ~@as(usize, 0)));

        if (@reduce(.Or, mask)) {
            return @intCast(u32, @reduce(.Min, min_indices));
        }

        indices += increment;        
    }

    if (remainder > 0) {
        i = slice.len - remainder;
        while (i < remainder): (i += 1) {
            const char = slice[i];
            if (char > max_u16_ascii[0] or char < min_u16_ascii[0]) {
                return @intCast(u32, i);
            }
        }
    }
    return null;
}

pub fn simd_single_loop(slice: []const u16) ?u32 {
    var remaining = slice;
    const end_ptr = remaining.ptr + remaining.len - (remaining.len % 8);
    if (remaining.len > 8) {
        const remaining_start = remaining.ptr;
        while (remaining.ptr != end_ptr) {
            const vec: @Vector(8, u16) = remaining[0..8].*;
            const max_value = @reduce(.Max, vec);

            // by using @reduce here, we make it only do one comparison
            // @reduce doesn't tell us the index though
            const min_value = @reduce(.Min, vec);
            if (min_value < 0x20 or max_value > 127) {
                remaining.len -= (@ptrToInt(remaining.ptr) - @ptrToInt(remaining_start)) / 2;

                // this is really slow
                // it does it element-wise for every single u8 on the vector
                // instead of doing the SIMD instructions
                // it removes a loop, but probably is slower in the end
                const cmp = @bitCast(@Vector(8, u1), vec > max_u16_ascii) |
                    @bitCast(@Vector(8, u1), vec < min_u16_ascii);
                const bitmask: u16 = @ptrCast(*const u16, &cmp).*;
                const first = @ctz(u16, bitmask);

                return @intCast(u32, @as(u32, first) +
                    @intCast(u32, slice.len - remaining.len));
            }
            remaining.ptr += 8;
        }
        remaining.len -= (@ptrToInt(remaining.ptr) - @ptrToInt(remaining_start)) / 2;
    }
    return null;
}

const std = @import("std");
pub fn main() anyerror!void {
    @setEvalBranchQuota(99999);
    const input = std.mem.span(toUTF16Literal("aspdokasdpokasdpokasd aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasd  aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasd aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasd aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123aspdokasdpokasdpokasdaspdokasdpokasdpokasdaspdokasdpokasdpokasd123123"));

    var timer = std.time.Timer.start() catch unreachable;
    _ = simd_double_loop(input);
    std.debug.print("simd Double Loop took {}\n", .{std.fmt.fmtDuration(timer.lap())});

    _ = simd_single_loop(input);
    std.debug.print("simd Single Loop took {}\n", .{std.fmt.fmtDuration(timer.lap())});
}
```

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
Thanks for this

If we read the source for `std.simd.firstIndexOfValue`:

```zig
const len = vectorLength(@TypeOf(vec));
    const IndexInt = VectorIndex(@TypeOf(vec));

    if (!@reduce(.Or, vec)) {
        return null;
    }
    const indices = @select(IndexInt, vec, iota(IndexInt, len), @splat(len, ~@as(IndexInt, 0)));
    return @reduce(.Min, indices);
```

We can see there's an extra `@reduce(.Or, vec)` in there

It will be faster if we use `@select` that way instead

**Comment by Jarred-Sumner** [comment]
~also, I think bun's zig version predates `std.simd`~ nope it doesn't

**Comment by Jarred-Sumner** [comment]
I tested it and turns out I'm wrong `std.simd` version is faster (unless I messed something up)

anyways, this is an improvement so I'm merging it regardless

File: src/string_immutable.zig [hunk 1]
```diff
@@ -3024,30 +3024,12 @@ pub fn firstNonASCII16CheckMin(comptime Slice: type, slice: Slice, comptime chec
                     // @reduce doesn't tell us the index though
                     const min_value = @reduce(.Min, vec);
                     if (min_value < 0x20 or max_value > 127) {
-                        remaining.len -= (@ptrToInt(remaining.ptr) - @ptrToInt(remaining_start)) / 2;
```
Comment (Jarred-Sumner) [inline src/string_immutable.zig:3064]: This line will need to be kept

**Comment by Jarred-Sumner** [comment]
on an M1, this improves the performance for UTF16 encoding/decoding but regresses for latin1

I think the reason why is the latin1 code uses SWAR instead of `@ctz` and SWAR seems pretty good, but harder to use for utf16

UTF16:
image

Latin1 (note the time spent on encodeInto)
image

**Comment by jonacollazo** [comment]
Hi @Jarred-Sumner I have been slowly working on this. Do you mind sharing the text-encoder test you used?
Would like to use it to test the regression for latin1

**Comment by gruckion** [comment]
The issue is marked as closed

https://github.com/oven-sh/bun/issues/188

Is this PR still needed?

**Comment by jdalton** [comment]
Closing for now. Please ping back if anything changes.

**Pull request closed without merging**